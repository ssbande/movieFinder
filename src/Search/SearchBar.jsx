import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { Form, Container } from 'semantic-ui-react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';

import Filters from './Filters';
import SearchField from './SearchField';
import Suggestions from './Suggestions';
import { getData, setFilter, setSubmitted, setUserQuery } from '../state/actions';
import { getFilter, getQuery, shouldShowSuggestions } from '../state/selectors';
import { HistoryProps, LocationProps, FilterProps } from '../utils/propTypes';

import './search.css';

const SearchBar = props => {
  const { filter, getData, setFilter, setUserQuery, showSuggestions, userQuery } = props;

  useEffect(() => {
    if (showSuggestions) getData(userQuery, true);
  }, [getData, showSuggestions, userQuery])

  const handleChange = ({ target: { value } }) => setUserQuery(value);
  const handleFilterChange = (evnt, { value }) => setFilter(value);
  
  const searchUserQuery = evnt => {
    evnt.preventDefault();

    const { history, location, setSubmitted } = props;
    setSubmitted(true);
    getData(userQuery);

    if (location.pathname !== "/") history.push('/')
  };

  return <Container text style={{ paddingTop: 30 }}>
    <Form inverted className='searchForm' onSubmit={searchUserQuery}>
      <Filters filter={filter} handleFilterChange={handleFilterChange} />
      <SearchField query={userQuery} handleChange={handleChange} />
      <Suggestions />
    </Form>
  </Container>
}

SearchBar.propTypes = {
  filter: FilterProps,
  getData: PropTypes.func.isRequired,
  history: HistoryProps,
  location: LocationProps,
  setUserQuery: PropTypes.func.isRequired,
  setFilter: PropTypes.func.isRequired,
  setSubmitted: PropTypes.func.isRequired,
  showSuggestions: PropTypes.bool.isRequired,
  userQuery: PropTypes.string
}

const mapDispatchToProps = {
  getData,
  setUserQuery,
  setFilter,
  setSubmitted,
};

const mapState = state => ({
  filter: getFilter(state),
  showSuggestions: shouldShowSuggestions(state),
  userQuery: getQuery(state),
})

export default withRouter(connect(mapState, mapDispatchToProps)(SearchBar));