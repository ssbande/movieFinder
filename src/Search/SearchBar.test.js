import React from 'react';
import { shallow, mount } from 'enzyme';

import SearchBar from './SearchBar';

const setup = () => ({
  props: {
    getData: jest.fn(),
    setUserQuery: jest.fn(),
    setFilter: jest.fn(),
    setSubmitted: jest.fn(),
    filter: 'all',
    showSuggestions: false,
    userQuery: 'searchString'
  }
})

describe('Search Bar', () => {
  it('should render correctly', () => {
    const { props } = setup();
    const wrapper = shallow(<SearchBar {...props} />);
    expect(wrapper).toMatchSnapshot();
  })

  it('should render correctly for showing suggestions', () => {
    const { props } = setup();
    props.showSuggestions = true;
    const wrapper = shallow(<SearchBar {...props} />);
    expect(wrapper).toMatchSnapshot();
  })

  it('should render correctly for different filters', () => {
    const { props } = setup();
    props.filter = 'movie';
    const wrapper = shallow(<SearchBar {...props} />);
    expect(wrapper).toMatchSnapshot();
  })
})