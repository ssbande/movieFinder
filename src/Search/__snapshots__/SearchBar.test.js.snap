// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Bar should render correctly 1`] = `ShallowWrapper {}`;

exports[`Search Bar should render correctly for different filters 1`] = `ShallowWrapper {}`;

exports[`Search Bar should render correctly for showing suggestions 1`] = `ShallowWrapper {}`;
